### ISSUE TITLE
Port Scalameta to Scala Native

### ISSUE THREAD

**Issue by DavidDudson** [scalameta/scalameta#772]
Given that SJS support was just added, It seems like SN support should be somewhat trivial now that we are not platform specific. I could be wrong, so lets see what actually needs to be done.

Here are the scalameta dependencies

- [x] scalatest
- [x] scalapb
- [x] utest
- [x] sourcecode
- [x] fastparse

Further tasks:

- [x] Compile testsNative
- [x] Link testsNative
- [x] Remove the s/CLASS/CLAZZ/g workaround (https://github.com/scala-native/scala-native/issues/1155)
- [x] Successfully run testsNative
- [x] Configure Travis to test native

**Comment by DavidDudson** [comment]
@densh I'm sure you can answer a lot of these questions I have.

**Comment by olafurpg** [comment]
This was one big motivation for porting to Scala.js/crossProject. I already tested the waters by trying to build scalafmt against scala-native, which resulted in these linking errors https://github.com/scala-native/scala-native/issues/557

**Comment by densh** [comment]
@olafurpg Most of the linking errors from that list are done now, we should retry and see if there is anything missing still.

**Comment by olafurpg** [comment]
@densh I just opened https://github.com/scalameta/scalafmt/pull/898 porting scalafmt-core to Scala.js. We can work from that branch to try build scalafmt with native, I don't suspect we won't hit on any linking errors in core, but maybe we need more work in scalafmt-cli.

I will try to look into porting to native tomorrow.

**Comment by olafurpg** [comment]
I am super excited to confirm that scalafmt links with scala-native 0.2!

![scalafmt-native](https://cloud.githubusercontent.com/assets/1408093/25481965/f4d2da18-2b4f-11e7-91df-125f818ab82e.gif)
 
Getting it to work only required removing a single `Enumeration` in the scalameta parser since it seems those use runtime reflection. The rest was simple build changes.

**Comment by xeno-by** [comment]
Amazing news! Can we get these changes in? Is the entire test suite
crosscompiling?

**Comment by olafurpg** [comment]
I didn't run the test suite, first we would need to get cross-built versions of these libraries released

- [ ] scalatest
- [ ] utest
- [ ] sourcecode
- [ ] fastparse
- [ ] scalacheck?

**Comment by xeno-by** [comment]
@densh Can you comment on what it would take to support these libraries?

**Comment by densh** [comment]
I have no data on scalatest as they have rather complicated build that's not easily migratable to native.

utest/sourcecode/fastparse seem to have enough APIs implemented in 0.2 already. We can't run the tests automatically from the sbt yet due to https://github.com/scala-native/scala-native/issues/339, but @Duhemm might have time to look at it in 0.3 milestone. 

Scalacheck should be either working already or require minimum changes from us to support it.

**Comment by olafurpg** [comment]
Fastparse 0.4.4 is out now with native support. Only remaining dependency is ScalaTest. Our options are to either

- don't run tests in native projects (booo!)
- write a thin FunSuite wrapper using utests.

I lean towards the second option. utest assert macro error messages are IME better than scalatest macro assert errors so I wouldn't mind switching to utest. Once https://github.com/lihaoyi/utest/pull/113 is merged then utest will also have pretty console reports.

**Comment by ShaneDelmore** [comment]
👍 for writing a FunSuite wrapper for uTest.

**Comment by xeno-by** [comment]
👍

**Comment by xeno-by** [comment]
I tried crosscompiling a small part of Scalameta to Native, and it worked pretty well, except for a minor inconvenienced described in https://github.com/scalameta/scalameta/pull/1228#issuecomment-359629104. All in all, this looks doable, and I'm looking forward to a native build of Scalatest to commence a full-scare migration.

**Comment by olafurpg** [comment]
scalatest 3.2 SNAP release is due out this week with native support! https://gitter.im/scalatest/scalatest?at=5a660edc517037a212dc18ee

**Comment by xeno-by** [comment]
\o/ Thank you, @bvenners!

**Comment by densh** [comment]
Issue with artifact names has been fixed by @sjrd (https://github.com/portable-scala/sbt-crossproject/pull/81) and is going to be part of the next release of sbt-crossproject.

**Comment by xeno-by** [comment]
@densh Thanks for the great news! Is there a gitter channel or something where you guys announce new releases?

**Comment by sjrd** [comment]
I announced the release on Twitter a few hours ago ;)

**Comment by xeno-by** [comment]
I started working on crosscompilation to native in https://github.com/xeno-by/scalameta/tree/topic/native. Here's a problem that I ran into:

```
> metapNative/run /Users/eburmako/Projects/scalameta/sandbox/META-INF/semanticdb/Library.semanticdb
[info] Compiling 1 protobuf files to /Users/eburmako/Projects/scalameta/semanticdb/semanticdb3/.native/target/scala-2.11/src_managed/main
[info] Compiling schema /Users/eburmako/Projects/scalameta/semanticdb/semanticdb3/semanticdb3.proto
protoc-jar: protoc version: 340, detected platform: mac os x/x86_64
protoc-jar: executing: [/var/folders/pg/v06175sd2_qb9jbg7c28xr040000gn/T/protocjar4860536607164199621/bin/protoc.exe, --plugin=protoc-gen-scala=/var/folders/pg/v06175sd2_qb9jbg7c28xr040000gn/T/protocbridge5139572597291050554, --scala_out=flat_package,grpc:/Users/eburmako/Projects/scalameta/semanticdb/semanticdb3/.native/target/scala-2.11/src_managed/main, -I/Users/eburmako/Projects/scalameta/semanticdb/semanticdb3, -I/Users/eburmako/Projects/scalameta/semanticdb/semanticdb3/.native/target/protobuf_external, /Users/eburmako/Projects/scalameta/semanticdb/semanticdb3/semanticdb3.proto]
[info] Compiling protobuf
[info] Protoc target directory: /Users/eburmako/Projects/scalameta/semanticdb/semanticdb3/.native/target/scala-2.11/src_managed/main
[info] Compiling 9 Scala sources to /Users/eburmako/Projects/scalameta/semanticdb/semanticdb3/.native/target/scala-2.11/classes...
[info] Compiling 1 Scala source to /Users/eburmako/Projects/scalameta/semanticdb/metap/.native/target/scala-2.11/classes...
[info] Linking (1780 ms)
[error] cannot link: @scala.meta.internal.semanticdb3.SymbolInformation$Kind$CLASS$
[error] cannot link: @scala.meta.internal.semanticdb3.SymbolInformation$Kind$class$
[error] cannot link: @scala.meta.internal.semanticdb3.SymbolInformation$Kind$class$::$init$_scala.meta.internal.semanticdb3.SymbolInformation$Kind_unit
[error] unable to link
[error] (metapNative/compile:nativeLinkNIR) unable to link
[error] Total time: 14 s, completed Feb 1, 2018 11:23:32 AM
```

@densh, what gives? Is this a native bug? If yes, is there a workaround that would allow me to ignore it and make progress regardless?

**Comment by xeno-by** [comment]
Here's another problem that I ran into. This one may be related to ScalaTest:

```
> testsNative/test
[info] Compiling 1 protobuf files to /Users/eburmako/Projects/scalameta/semanticdb/semanticdb3/.native/target/scala-2.11/src_managed/main
...
[info] Compiling 1 Scala source to /Users/eburmako/Projects/scalameta/tests/native/target/scala-2.11/test-classes...
[error] /Users/eburmako/Projects/scalameta/tests/native/target/scala-2.11/src_managed/nativetest/FrameworksMap.scala:4: type ErrorSuite is not a member of package <root>
[error]   override val tests = Map[String, AnyRef]("scala.meta.tests.contrib.StatExtractionTest" -> new _root_.scala.meta.tests.contrib.StatExtractionTest, ...
[error] one error found
[error] (testsNative/nativetest:compileIncremental) Compilation failed
[error] Total time: 129 s, completed Feb 1, 2018 11:42:59 AM
```

@bvenners, any idea what could be causing the issue?

**Comment by shadaj** [comment]
Looks like https://github.com/scala-native/scala-native/blob/17cf3c557a10b63898b2fa4251b8a298177716df/sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/TestUtilities.scala#L38 is generating the tests map with `_root_` prepended to every test class, which leads to the failure since `_root_` can't be used to access values or classes in the root package.

@Duhemm would it be possible to generate the `TestMain` into the root package so that `_root_` is not needed to reference test classes/modules?

**Comment by xeno-by** [comment]
I've just submitted a pull request with the first stab at crosscompiling Scalameta to Scala Native: https://github.com/scalameta/scalameta/pull/1243. I had to apply several workarounds, but I managed to get testsNative/test to green state. If you're interested, check out the pull request for details.

**Comment by xeno-by** [comment]
Here's another pull request that crosscompiles just `semanticdb/`: https://github.com/scalameta/scalameta/pull/1263, but without any workarounds and hacks.

**Comment by xeno-by** [comment]
Successfully merged #1263 and published a Scala Native build of `semanticdb3`. Already using it in Rsc: https://github.com/xeno-by/reasonable-scala/commit/51f44e2ce81627fedc0738c02bf2595b90740737.

**Comment by olafurpg** [comment]
Fixed in #1529, native support for scalameta + contrib modules has been merged into master 🚀
